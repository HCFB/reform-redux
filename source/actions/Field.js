import {
  CHANGE_FIELD_VALUE,
  SET_FIELD_ERRORS,
  SET_FIELDS_ERRORS,
  SET_FIELD_DISABLED,
  SET_FIELDS_DISABLED,
  CHANGE_FIELDS_VALUES,
  RESET_FIELD,
  RESET_FIELDS,
  REMOVE_FIELD,
  SET_FIELD_TOUCHED,
  SET_FIELDS_TOUCHED,
} from '../constants/Field';
import type {
  ChangeFieldValue,
  SetFieldErrors,
  SetFieldsErrors,
  SetFieldsDisabled,
  SetFieldDisabled,
  ChangeFieldsValues,
  ResetField,
  ResetFields,
  FieldValue,
  FieldName,
  RemoveField,
  SetFieldTouched,
  SetFieldsTouched,
} from '../types/Field';
import { getReduxConst } from '../utils/common';

export const setFieldTouched: Function = (
  formName: string,
  fieldName: FieldName,
  fieldTouched: boolean,
): SetFieldTouched => ({
  type: getReduxConst(SET_FIELD_TOUCHED),
  formName,
  fieldName,
  fieldTouched,
});

export const setFieldsTouched: Function = (
  formName: string,
  touchedFields: { [fieldName: FieldName]: boolean },
): SetFieldsTouched => ({
  type: getReduxConst(SET_FIELDS_TOUCHED),
  formName,
  touchedFields,
});

export const removeField: Function = (formName: string, fieldName: FieldName): RemoveField => ({
  type: getReduxConst(REMOVE_FIELD),
  formName,
  fieldName,
});

export const resetField: Function = (formName: string, fieldName: FieldName): ResetField => ({
  type: getReduxConst(RESET_FIELD),
  formName,
  fieldName,
});

export const resetFields: Function = (
  formName: string,
  fieldsNames: Array<FieldName>,
): ResetFields => ({
  type: getReduxConst(RESET_FIELDS),
  formName,
  fieldsNames,
});

export const setFieldDisabled: Function = (
  formName: string,
  fieldName: FieldName,
  disabled: boolean = true,
): SetFieldDisabled => ({
  type: getReduxConst(SET_FIELD_DISABLED),
  formName,
  fieldName,
  disabled,
});

export const setFieldsDisabled: Function = (
  formName: string,
  disabledFields: { [fieldName: FieldName]: boolean },
): SetFieldsDisabled => ({
  type: getReduxConst(SET_FIELDS_DISABLED),
  formName,
  disabledFields,
});

export const changeFieldValue: Function = (
  formName: string,
  fieldName: FieldName,
  fieldValue: FieldValue,
): ChangeFieldValue => ({
  type: getReduxConst(CHANGE_FIELD_VALUE),
  formName,
  fieldName,
  fieldValue,
});

export const changeFieldsValues: Function = (
  formName: string,
  fieldsValues: { [fieldName: FieldName]: FieldValue },
): ChangeFieldsValues => ({
  type: getReduxConst(CHANGE_FIELDS_VALUES),
  formName,
  fieldsValues,
});

export const setFieldErrors: Function = (
  formName: string,
  fieldName: FieldName,
  errors: Array<string>,
): SetFieldErrors => ({
  type: getReduxConst(SET_FIELD_ERRORS),
  formName,
  fieldName,
  errors,
});

export const setFieldsErrors: Function = (
  formName: string,
  fieldsErrors: { [fieldName: FieldName]: Array<string> },
): SetFieldsErrors => ({
  type: getReduxConst(SET_FIELDS_ERRORS),
  formName,
  fieldsErrors,
});
