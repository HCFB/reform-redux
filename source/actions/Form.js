import {
  FORM_INITIALISATION,
  RESET_FORM,
  SET_FORM_SUBMITTING,
  UPDATE_FORM,
  SET_FORM_SUBMITTED,
} from '../constants/Form';
import type { ResetState } from '../types/formReducer';
import { getReduxConst } from '../utils/common';
import type { FieldsData } from '../types/Field';
import type {
  FormInitialisation,
  SetFormSubmitting,
  ResetForm,
  UpdateForm,
  SetFormSubmitted,
} from '../types/Form';

export const formInitialisation = (formName: string, fields: FieldsData): FormInitialisation => ({
  type: getReduxConst(FORM_INITIALISATION),
  formName,
  fields,
});

export const updateForm = (formName: string, fields: FieldsData): UpdateForm => ({
  type: getReduxConst(UPDATE_FORM),
  formName,
  fields,
});

export const setFormSubmitting = (formName: string, submitting: boolean): SetFormSubmitting => ({
  type: getReduxConst(SET_FORM_SUBMITTING),
  formName,
  submitting,
});

/**
 * Set form submitted
 *
 * @example
 *
 * import { setFormSubmitted } from 'reform-redux';
 *
 * store.dispatch(setFormSubmitted(
 *   formName: string,
 *   submitted: boolean,
 * ));
 *
 * @callback
 * @param {string} formName
 * @param {boolean} submitted
 * @returns {SetFormSubmitted}
 */
export const setFormSubmitted = (formName: string, submitted: boolean): SetFormSubmitted => ({
  type: getReduxConst(SET_FORM_SUBMITTED),
  formName,
  submitted,
});

/**
 * Reset form
 *
 * @example
 *
 * import { resetForm } from 'reform-redux';
 *
 * store.dispatch(resetForm(
 *   formName: string,
 *   state?: ResetState,
 * ));
 *
 * @callback
 * @param {string} formName
 * @param {ResetState} [state]
 */
export const resetForm = (formName: string, state?: ResetState): ResetForm => ({
  type: getReduxConst(RESET_FORM),
  formName,
  state,
});
