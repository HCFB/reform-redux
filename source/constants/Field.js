export const CHANGE_FIELD_VALUE: string = 'CHANGE_FIELD_VALUE';
export const CHANGE_FIELDS_VALUES: string = 'CHANGE_FIELDS_VALUES';
export const SET_FIELD_ERRORS: string = 'SET_FIELD_ERRORS';
export const SET_FIELDS_ERRORS: string = 'SET_FIELDS_ERRORS';
export const SET_FIELD_DISABLED: string = 'SET_FIELD_DISABLED';
export const SET_FIELDS_DISABLED: string = 'SET_FIELDS_DISABLED';
export const RESET_FIELD: string = 'RESET_FIELD';
export const RESET_FIELDS: string = 'RESET_FIELDS';
export const REMOVE_FIELD: string = 'REMOVE_FIELD';
export const SET_FIELD_TOUCHED: string = 'SET_FIELD_TOUCHED';
export const SET_FIELDS_TOUCHED: string = 'SET_FIELDS_TOUCHED';
export const SET_FIELD_CHANGED: string = 'SET_FIELD_CHANGED';
export const SET_FIELDS_CHANGED: string = 'SET_FIELDS_CHANGED';
