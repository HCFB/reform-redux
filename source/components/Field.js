import { Component, createElement } from 'react';
import PropTypes from 'prop-types';
import { getValidateFunctionsArray, validateField } from '../utils/Field';
import type { MiniReduxForm } from '../types/Form';
import type { FieldData, FieldsData, ComponentProps, ComponentState } from '../types/Field';
import type { State } from '../types/formReducer';
import type { DataFunctions } from '../types/dataFunctions';
import type { ComponentCreator } from '../types/common';

export const createFieldComponent: ComponentCreator = (dataFunctions: DataFunctions) => {
  const {
    getIn,
    map,
    list,
    setIn,
    is,
    isList,
    listIncludes,
    isImmutable,
    toJS,
    keys,
    listSize,
  }: DataFunctions = dataFunctions;

  class Field extends Component<ComponentProps, ComponentState> {
    initialFieldData: FieldData;
    fieldWasTouched: boolean = false;
    unsubscribeFromStore: Function = () => {};
    reduxRenderCount: number = 0;

    static defaultProps: {
      disabled: $PropertyType<ComponentProps, 'disabled'>,
      value: $PropertyType<ComponentProps, 'value'>,
    } = {
      value: '',
      disabled: false,
    };

    static contextTypes = {
      _reformRedux: PropTypes.object,
      store: PropTypes.object,
    };

    constructor(props: ComponentProps, context: MiniReduxForm) {
      super(props, context);

      if (!context._reformRedux) {
        throw new Error('Component `Field` must be in `Form` component.');
      }

      if (!props.name) {
        throw new Error('The `name` prop is required.');
      }

      if (props.normalize && typeof props.normalize !== 'function') {
        throw new Error('The `normalize` prop must be a function.');
      }

      if (!props.component) {
        throw new Error('The `component` prop is required.');
      }

      if (props.multiple && props.component === 'select' && props.value && !isList(props.value)) {
        throw new Error(
          'The `value` prop supplied to Field with type "select" must be an array if `multiple` is true.',
        );
      }

      // Normalize value on component onInit

      let value = props.value || '';

      if (props.normalize) {
        const { normalize } = props;
        value = normalize(props.value, '', map({}), 'onInit');
      }

      this.initialFieldData = map({
        value,
        errors: list([]),
        valid: true,
        disabled: props.disabled || false,
      });

      if (['radio', 'checkbox'].indexOf(props.type) > -1 && !props.checked) {
        this.initialFieldData = setIn(this.initialFieldData, ['value'], '');
      }

      // Get default value from store if it exists

      const state: State = this.context.store.getState();
      const currentFormData: State = getIn(state, this.context._reformRedux.form.path);
      const initialFieldData: FieldData = getIn(currentFormData, ['fields', props.name]);

      if (initialFieldData) {
        this.initialFieldData = initialFieldData;
      }

      // Write initial data to store

      this.state = {
        field: map(this.initialFieldData),
      };
    }

    shouldComponentUpdate() {
      return Boolean(this.reduxRenderCount > 1);
    }

    componentWillMount() {
      this.unsubscribeFromStore = this.context.store.subscribe(() => {
        const state: State = this.context.store.getState();
        const currentFormData: State = getIn(state, this.context._reformRedux.form.path);
        const currentFieldData: FieldData = this.state.field;
        const nextFieldData: { field: FieldData } = {
          field: getIn(
            state,
            [...this.context._reformRedux.form.path, 'fields', this.props.name],
            this.state.field,
          ),
        };

        if (currentFormData.submitted) {
          this.fieldWasTouched = true;
        }

        this.reduxRenderCount += 1;

        if (!is(currentFieldData, nextFieldData.field)) {
          this.setState({
            field: map(nextFieldData.field),
          });
        }
      });

      const initialFieldData = map(this.initialFieldData);
      const { type, checked, multiple, component } = this.props;

      const validate: Array<Function> = getValidateFunctionsArray(dataFunctions)(
        this.props.validate,
      );
      const state: State = this.context.store.getState();
      const currentFormData: State = getIn(state, this.context._reformRedux.form.path);

      this.context._reformRedux.form.registerField(this.props.name, initialFieldData, validate, {
        type,
        checked,
        multiple,
        component,
      });

      if (listSize(keys(getIn(currentFormData, ['fields'])))) {
        this.context._reformRedux.form.updateForm();
      }
    }

    componentWillUnmount() {
      this.unsubscribeFromStore();
      this.context._reformRedux.form.unregisterField(this.props.name);
    }

    componentWillReceiveProps(nextProps: ComponentProps) {
      if (nextProps.value !== undefined && !is(map(this.props.value), map(nextProps.value))) {
        this.changeFieldValue(nextProps.value);
      }
    }

    setFieldErrors = (errors: Array<string>) => {
      if (!is(getIn(this.state.field, ['errors']), errors)) {
        this.context._reformRedux.field.setFieldErrors(this.props.name, errors);
      }
    };

    changeFieldValue = async (value: any) => {
      this.context._reformRedux.field.changeFieldValue(this.props.name, value);

      if (this.fieldWasTouched && this.props.validate) {
        const validate: Array<Function> = getValidateFunctionsArray(dataFunctions)(
          this.props.validate,
        );

        const errors: Array<string> = await validateField(value, validate, dataFunctions);

        this.setFieldErrors(errors);
      }
    };

    getFieldValue = (data: any): any => {
      if (data.nativeEvent && data.nativeEvent instanceof Event) {
        if (this.isRadio()) {
          return data.target.checked ? this.props.value : '';
        }

        if (this.isCheckbox()) {
          if (this.context._reformRedux.form.fieldsCount[this.props.name] > 1) {
            return list(
              data.target.checked
                ? [...toJS(getIn(this.state.field, ['value'])), toJS(this.props.value)]
                : toJS(getIn(this.state.field, ['value'])).filter(
                  value => value !== this.props.value, // eslint-disable-line
                ), // eslint-disable-line
            );
          }

          return data.target.checked ? this.props.value : '';
        }

        if (this.props.component === 'select' && this.props.multiple) {
          return list(
            [].slice.call(data.target.selectedOptions).map(option => {
              return option.value;
            }),
          );
        }

        return data.target.value;
      }

      return data;
    };

    changeFieldValueHandler = (data: any, normalizeWhen: string = 'onChange') => {
      let value: any = this.getFieldValue(data);
      const { normalize } = this.props;

      if (normalize) {
        const state: State = this.context.store.getState();
        const currentFormData: State = getIn(state, this.context._reformRedux.form.path);
        const fields: FieldsData = getIn(currentFormData, ['fields']);

        value = normalize(value, getIn(this.state.field, ['value']), map(fields), normalizeWhen);
      }

      this.changeFieldValue(value);

      if (this.props.onChange) {
        this.props.onChange(data, value);
      }
    };

    onBlurFieldHandler = async (event: Event) => {
      if (this.props.onBlur) {
        const fieldData: FieldData = this.state.field;
        this.props.onBlur(event, fieldData);
      }

      const fieldValue: any = getIn(this.state.field, ['value']);

      if (this.props.normalize) {
        this.changeFieldValueHandler(fieldValue, 'onBlur');
      }

      // If the field was touched don't validate him.

      if (this.fieldWasTouched) return;
      this.fieldWasTouched = true;

      if (this.props.validate) {
        const validate: Array<Function> = getValidateFunctionsArray(dataFunctions)(
          this.props.validate,
        );

        const errors: Array<string> = await validateField(fieldValue, validate, dataFunctions);

        this.setFieldErrors(errors);
      }
    };

    onFocusFieldHandler = (event: Event) => {
      if (this.props.normalize) {
        this.changeFieldValueHandler(getIn(this.state.field, ['value']), 'onFocus');
      }

      if (this.props.onFocus) {
        const fieldData: FieldData = this.state.field;
        this.props.onFocus(event, fieldData);
      }
    };

    isRadioOrCheckbox = () => {
      return this.isRadio() || this.isCheckbox();
    };

    isCheckbox = () => {
      const { type } = this.props;
      return type && type === 'checkbox';
    };

    isRadio = () => {
      const { type } = this.props;
      return type && type === 'radio';
    };

    render() {
      // eslint-disable-next-line no-unused-vars
      const { normalize, component, name, validate, disabled, ...restProps } = this.props;
      let fieldProps: {
        onChange: Function,
        onBlur: Function,
        onFocus: Function,
        value: any,
        disabled: boolean,
      } = {
        ...restProps,
        onChange: this.changeFieldValueHandler,
        onBlur: this.onBlurFieldHandler,
        onFocus: this.onFocusFieldHandler,
        value: getIn(this.state.field, ['value']),
        disabled: getIn(this.state.field, ['disabled']),
      };

      const fieldValue: any = getIn(this.state.field, ['value']);

      if (this.isRadioOrCheckbox()) {
        fieldProps = {
          ...fieldProps,
          checked: is(restProps.value, fieldValue),
          value: restProps.value,
        };
      }

      if (this.isCheckbox() && isList(fieldValue)) {
        fieldProps = {
          ...fieldProps,
          checked: listIncludes(fieldValue, restProps.value),
        };
      }

      if (typeof component !== 'string') {
        fieldProps = {
          ...fieldProps,
          formName: this.context._reformRedux.form.name,
          errors: getIn(this.state.field, ['errors']),
        };
      } else {
        fieldProps = {
          ...fieldProps,
          value: isImmutable(fieldValue) ? toJS(fieldValue) : fieldValue,
        };
      }

      if (component === 'select' && restProps.multiple && !fieldValue) {
        fieldProps = {
          ...fieldProps,
          value: [],
        };
      }

      return createElement(this.props.component, fieldProps);
    }
  }

  return Field;
};
