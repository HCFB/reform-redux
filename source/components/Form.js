import { Component, createElement } from 'react';
import PropTypes from 'prop-types';
import {
  formInitialisation,
  resetForm,
  setFormSubmitting,
  updateForm,
  setFormSubmitted,
} from '../actions/Form';
import {
  changeFieldValue,
  setFieldErrors,
  setFieldsErrors,
  changeFieldsValues,
  setFieldDisabled,
  setFieldsDisabled,
  resetField,
  resetFields,
  removeField,
  setFieldsTouched,
  setFieldTouched,
  setFieldChanged,
  setFieldsChanged,
} from '../actions/Field';
import { validateField, getValidateFunctionsArray } from '../utils/Field';
import { debounce, asyncForEach, filterReactDomProps } from '../utils/common';
import type { Element } from 'react';
import type { ReFormRedux, ComponentProps, FieldsValidate } from '../types/Form';
import type {
  FieldData,
  FieldsData,
  FieldValidateProp,
  FieldsCount,
  FieldName,
} from '../types/Field';
import type { ComponentCreator } from '../types/common';
import type { Store } from 'redux';
import type { State, ResetState } from '../types/formReducer';
import type { DataFunctions } from '../types/dataFunctions';

export const createFormComponent: ComponentCreator = (dataFunctions: DataFunctions) => {
  const {
    getIn,
    keys,
    listSize,
    list,
    setIn,
    map,
    isList,
    deleteIn,
  }: DataFunctions = dataFunctions;

  class Form extends Component<ComponentProps> {
    formName: string;
    path: Array<string>;
    initialized: boolean = false;

    static contextTypes = {
      store: PropTypes.object,
    };

    static childContextTypes = {
      _reformRedux: PropTypes.object,
    };

    static propTypes = {
      path: PropTypes.string.isRequired,
      name: PropTypes.string,
    };

    static defaultProps: {
      onSubmit: $PropertyType<ComponentProps, 'onSubmit'>,
      onSubmitFailed: $PropertyType<ComponentProps, 'onSubmitFailed'>,
    } = {
      onSubmit: () => {},
      onSubmitFailed: () => {},
    };

    fieldsStack: {
      [formName: string]: FieldsData,
    } = {};
    fieldsCount: {
      [formName: string]: FieldsCount,
    } = {};
    fieldsValidateStack: {
      [formName: string]: FieldsValidate,
    } = {};

    constructor(props: ComponentProps, context) {
      super(props);

      if (!props.path) {
        throw new Error('The `path` prop is required.');
      }

      this.path = props.path.split('.');
      this.formName = props.name || this.path.slice(-1)[0];

      if (!this.fieldsStack[this.formName]) this.fieldsStack[this.formName] = {};
      if (!this.fieldsCount[this.formName]) this.fieldsCount[this.formName] = {};
      if (!this.fieldsValidateStack[this.formName]) this.fieldsValidateStack[this.formName] = {};

      this.updateForm = this.createFormUpdater(context.store);
    }

    getChildContext(): ReFormRedux {
      const store: Store<State, *, *> = this.context.store;

      return {
        _reformRedux: {
          form: {
            name: this.formName,
            path: this.path,
            fieldsCount: this.fieldsCount[this.formName],
            registerField: this.registerField,
            unregisterField: this.unregisterField,
            resetForm: (state?: ResetState): Function =>
              store.dispatch(resetForm(this.formName, state)),
            setFormSubmitted: (submitted: boolean): Function =>
              store.dispatch(setFormSubmitted(this.formName, submitted)),
          },
          field: {
            setFieldTouched: (fieldName: FieldName, fieldTouched: boolean): Function =>
              store.dispatch(setFieldTouched(this.formName, fieldName, fieldTouched)),
            setFieldsTouched: (fieldsTouched: { [fieldName: FieldName]: boolean }): Function =>
              store.dispatch(setFieldsTouched(this.formName, fieldsTouched)),
            setFieldChanged: (fieldName: FieldName, fieldChanged: boolean): Function =>
              store.dispatch(setFieldChanged(this.formName, fieldName, fieldChanged)),
            setFieldsChanged: (fieldsChanged: { [fieldName: FieldName]: boolean }): Function =>
              store.dispatch(setFieldsChanged(this.formName, fieldsChanged)),
            removeField: (fieldName: FieldName): Function =>
              store.dispatch(removeField(this.formName, fieldName)),
            changeFieldsValues: (fieldsValues: { [fieldName: FieldName]: any }): Function =>
              store.dispatch(changeFieldsValues(this.formName, fieldsValues)),
            changeFieldValue: (fieldName: FieldName, fieldValue: any): Function =>
              store.dispatch(changeFieldValue(this.formName, fieldName, fieldValue)),
            setFieldErrors: (fieldName: FieldName, errors: Array<string>): Function =>
              store.dispatch(setFieldErrors(this.formName, fieldName, errors)),
            setFieldsErrors: (
              fieldName: FieldName,
              fieldsErrors: { [fieldName: FieldName]: Array<string> },
            ): Function => store.dispatch(setFieldsErrors(this.formName, fieldsErrors)),
            setFieldDisabled: (fieldName: FieldName, disabled: boolean = true): Function =>
              store.dispatch(setFieldDisabled(this.formName, fieldName, disabled)),
            setFieldsDisabled: (disabledFields: { [fieldName: FieldName]: boolean }): Function =>
              store.dispatch(setFieldsDisabled(this.formName, disabledFields)),
            resetField: (fieldName: FieldName, state?: ResetState): Function =>
              store.dispatch(resetField(this.formName, fieldName, state)),
            resetFields: (fieldsNames: Array<FieldName>, state?: ResetState): Function =>
              store.dispatch(resetFields(this.formName, fieldsNames, state)),
          },
        },
      };
    }

    createFormUpdater = (store: Store<State, *, *>) =>
      debounce(
        (): Function => store.dispatch(updateForm(this.formName, this.fieldsStack[this.formName])),
        250,
      );

    increaseFieldCount = (fieldName: FieldName) => {
      const fieldsCount: number = this.fieldsCount[this.formName][fieldName] || 0;
      return (this.fieldsCount[this.formName][fieldName] = fieldsCount + 1);
    };

    decreaseFieldCount = (fieldName: FieldName) => {
      const fieldsCount: number = this.fieldsCount[this.formName][fieldName];
      return (this.fieldsCount[this.formName][fieldName] = fieldsCount ? fieldsCount - 1 : 0);
    };

    unregisterField = (fieldName: FieldName, removeOnUnmount: boolean) => {
      this.decreaseFieldCount(fieldName);

      if (removeOnUnmount) {
        this.context.store.dispatch(removeField(this.formName, fieldName));
        this.fieldsStack = deleteIn(this.fieldsStack, [this.formName, fieldName]);
        this.fieldsValidateStack = deleteIn(this.fieldsValidateStack, [this.formName, fieldName]);
      }
    };

    registerField = (
      fieldName: FieldName,
      fieldData: FieldData,
      fieldValidate: FieldValidateProp,
      fieldAdditionalData: {
        type: string,
        checked: boolean,
        multiple: boolean,
        component: string,
      },
    ) => {
      this.increaseFieldCount(fieldName);

      if (fieldAdditionalData.type && this.fieldsCount[this.formName][fieldName] > 1) {
        if (fieldAdditionalData.type === 'radio' && !fieldAdditionalData.checked) {
          return;
        }

        if (fieldAdditionalData.type === 'checkbox' || fieldAdditionalData.type === 'radio') {
          if (fieldAdditionalData.checked) {
            if (fieldAdditionalData.type === 'checkbox') {
              if (!isList(getIn(this.fieldsStack[this.formName][fieldName], ['value']))) {
                let fieldValue: any = list([getIn(fieldData, ['value'])]);

                if (getIn(this.fieldsStack[this.formName][fieldName], ['value'])) {
                  fieldValue = list([
                    getIn(this.fieldsStack[this.formName][fieldName], ['value']),
                    getIn(fieldData, ['value']),
                  ]);
                }

                return (this.fieldsStack[this.formName][fieldName] = setIn(
                  this.fieldsStack[this.formName][fieldName],
                  ['value'],
                  fieldValue,
                ));
              }

              return (this.fieldsStack[this.formName][fieldName] = setIn(
                this.fieldsStack[this.formName][fieldName],
                ['value', listSize(getIn(this.fieldsStack[this.formName][fieldName], ['value']))],
                getIn(fieldData, ['value']),
              ));
            }

            return (this.fieldsStack[this.formName][fieldName] = setIn(
              this.fieldsStack[this.formName][fieldName],
              ['value'],
              getIn(fieldData, ['value']),
            ));
          }

          if (
            !isList(getIn(this.fieldsStack[this.formName][fieldName], ['value'])) &&
            fieldAdditionalData.type === 'checkbox'
          ) {
            let fieldValue: any = list([]);

            if (getIn(this.fieldsStack[this.formName][fieldName], ['value'])) {
              fieldValue = list([getIn(this.fieldsStack[this.formName][fieldName], ['value'])]);
            }

            return (this.fieldsStack[this.formName][fieldName] = setIn(
              this.fieldsStack[this.formName][fieldName],
              ['value'],
              fieldValue,
            ));
          }

          return;
        }
      }

      // Set empty array as default value for multiple select

      if (
        fieldAdditionalData.component === 'select' &&
        fieldAdditionalData.multiple &&
        !getIn(fieldData, ['value'])
      ) {
        fieldData = setIn(fieldData, ['value'], list([]));
      }

      this.fieldsStack[this.formName][fieldName] = fieldData;
      this.fieldsValidateStack[this.formName][fieldName] = fieldValidate;

      if (this.initialized) {
        this.updateForm();
      }
    };

    componentDidMount() {
      const state: State = this.context.store.getState();
      const currentFormData = getIn(state, this.path);
      const fields: FieldsData = getIn(currentFormData, ['fields']);

      if (!listSize(keys(fields))) {
        this.context.store.dispatch(
          formInitialisation(this.formName, this.fieldsStack[this.formName]),
        );
        this.initialized = true;
      }
    }

    handleSubmit = async (event: Event) => {
      event.preventDefault();

      const store: Store<State, *, *> = this.context.store;
      store.dispatch(setFormSubmitting(this.formName, true));

      const { onSubmit, onSubmitFailed } = this.props;

      // Validate all fields

      let state: State = store.getState();
      let fields: FieldsData = getIn(state, [...this.path, 'fields']);
      let fieldsErrors: { [fieldName: FieldName]: Array<string> } = map({});
      let errorsExists: boolean = false;

      await asyncForEach(
        keys(fields),
        async (fieldKey: string) => {
          const validateFunctions = getValidateFunctionsArray(dataFunctions)(
            this.fieldsValidateStack[this.formName][fieldKey],
          );
          let errors: Array<string> = getIn(fields, [fieldKey, 'errors']);

          if (!listSize(errors)) {
            errors = await validateField(
              getIn(fields, [fieldKey, 'value']),
              validateFunctions,
              dataFunctions,
            );
          }

          fieldsErrors = setIn(fieldsErrors, [fieldKey], errors);

          if (listSize(errors)) {
            errorsExists = true;
          }
        },
        dataFunctions,
      );

      if (errorsExists) {
        store.dispatch(setFieldsErrors(this.formName, fieldsErrors));

        state = store.getState();
        fields = getIn(state, [...this.path, 'fields']);

        let fieldsWithErrors: { [fieldName: FieldName]: FieldData } = map({});

        keys(fields).forEach((fieldKey: FieldName) => {
          if (!getIn(fields, [fieldKey, 'valid'])) {
            fieldsWithErrors = setIn(fieldsWithErrors, [fieldKey], getIn(fields, [fieldKey]));
          }
        });

        if (onSubmitFailed) {
          onSubmitFailed(fieldsWithErrors, fields, event);
          store.dispatch(setFormSubmitting(this.formName, false));
        }
      } else if (onSubmit) {
        state = store.getState();
        fields = getIn(state, [...this.path, 'fields']);

        Promise.resolve(onSubmit(fields, event)).then(() => {
          store.dispatch(setFormSubmitting(this.formName, false));
          store.dispatch(setFormSubmitted(this.formName, true));
        });
      }
    };

    render(): Element<'form'> {
      const { children } = this.props;

      return createElement('form', {
        ...filterReactDomProps(this.props),
        onSubmit: this.handleSubmit,
        children,
      });
    }
  }

  return Form;
};
