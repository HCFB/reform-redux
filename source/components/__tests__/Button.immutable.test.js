import { createElement } from 'react';
import { getIn } from 'immutable';
import { shallow, mount } from 'enzyme';
import { Button } from '../../immutable';
import { formInitialisation, setFormSubmitting } from '../../actions/Form';
import { changeFieldValue } from '../../actions/Field';

describe('components / Button', () => {
  it('snapshot', () => {
    const snapshot = shallow(
      createElement(global.Provider, { immutable: true }, createElement(Button)),
    );

    expect(snapshot).toMatchSnapshot();
  });

  it('if component is not in Form component then throw error', () => {
    expect(() => shallow(createElement(Button))).toThrow(
      'Component `Button` must be in `Form` component.',
    );
  });

  it('component with type=reset will reset form on onClick', () => {
    const component = mount(
      createElement(global.Provider, { immutable: true }, createElement(Button, { type: 'reset' })),
    );

    global.immutableStore.dispatch(
      formInitialisation('form', {
        field: {
          value: '',
          errors: [],
          valid: true,
          disabled: false,
        },
      }),
    );

    global.immutableStore.dispatch(changeFieldValue('form', 'field', 'test'));
    component.find('button').simulate('click');

    expect(getIn(global.immutableStore.getState(), ['form', 'fields', 'field', 'value'])).toBe('');
  });

  it('component with type=submit will just trigger onClick', () => {
    expect.assertions(2);

    const onClick = jest.fn();
    const component = mount(
      createElement(
        global.Provider,
        { immutable: true },
        createElement(Button, { type: 'submit', onClick }),
      ),
    );

    global.immutableStore.dispatch(
      formInitialisation('form', {
        field: {
          value: '',
          errors: [],
          valid: true,
          disabled: false,
        },
      }),
    );

    global.immutableStore.dispatch(changeFieldValue('form', 'field', 'test'));

    component.find('button').simulate('click');

    expect(getIn(global.immutableStore.getState(), ['form', 'fields', 'field', 'value'])).not.toBe(
      '',
    );
    expect(onClick).toBeCalledWith(expect.anything());
  });

  it('disabled property', () => {
    const component = mount(
      createElement(
        global.Provider,
        { immutable: true },
        createElement(Button, { disabled: true }),
      ),
    );
    expect(component.find('button').props().disabled).toBeTruthy();
  });

  it('after change form property "submitting", component property "disabled" and state "submitting" will the same', () => {
    expect.assertions(2);

    const component = mount(
      createElement(global.Provider, { immutable: true }, createElement(Button)),
    );
    let buttonComponent = component.find('Button').instance();
    let button = component.find('button');

    expect(button.prop('disabled') && buttonComponent.state.submitting).toBeFalsy();

    global.immutableStore.dispatch(setFormSubmitting('form', true));
    component.update();
    buttonComponent = component.find('Button').instance();
    button = component.find('button');

    expect(button.prop('disabled') && buttonComponent.state.submitting).toBeTruthy();
  });

  it('if component was unmounted then after change form property "submitting", component property "disabled" and state "submitting" will not the same', () => {
    const component = mount(createElement(Button), {
      context: global.immutableContext,
    });
    component.unmount();

    global.immutableStore.dispatch(setFormSubmitting('form', true));

    component.mount();

    expect(component.prop('disabled') && component.state('submitting')).toBeFalsy();
  });
});
