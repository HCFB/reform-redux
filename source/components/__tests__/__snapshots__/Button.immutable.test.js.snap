// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components / Button.immutable snapshot 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Provider
    immutable={true}
  >
    <ForwardRef />
  </Provider>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": <ForwardRef
        path="form"
      >
        <ForwardRef />
      </ForwardRef>,
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": <ForwardRef />,
        "path": "form",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "render": [Function],
      },
    },
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": <ForwardRef
          path="form"
        >
          <ForwardRef />
        </ForwardRef>,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": <ForwardRef />,
          "path": "form",
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": Object {
            "$$typeof": Symbol(react.forward_ref),
            "render": [Function],
          },
        },
        "type": Object {
          "$$typeof": Symbol(react.forward_ref),
          "render": [Function],
        },
      },
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
